### ISSUE TITLE
Create rule XXX: Modules should be stateless / Modules should not be a hidden singleton

### ISSUE THREAD

**Issue by yassin-kammoun-sonarsource** [SonarSource/SonarJS#4418]
#### Detection

- A module should not have mutable variables, that is, variable declarations with `let` at the top-level scope

#### Noncompliant

```js
let x = 42; // Noncompliant

export function foo(y) {
  x += y;
  return x;
}
```

#### Compliant

```js
export default (function () {
    let instance;
 
    function createInstance() {
        let object = {
           x: 42,
           foo(y) {
              this.x += y;
              return this.x;
           }
        }
        return object;
    }

    return {
        getInstance: function () {
            if (!instance) {
                instance = createInstance();
            }
            return instance;
        }
    };
})();
```

**Comment by gabriel-vivas-sonarsource** [comment]
The example we provide is no more testable than the original IMO.

Alternatively, one could export both the constructor as a named export and an instance as the default. that would be easier to test in isolation, but arguably still a bad pattern because it gets complicated when the module is deep in a hierarchy.

I think the better compliant version would be to export the constructor function and use it in a different module.

**Comment by ilia-kebets-sonarsource** [comment]
As discussed, we could also flag objects and arrays whose reference is constant, but that are mutated like:
```js
const stack = [];
export function foo(num) {
  stack.push(num);
}
```

**Comment by Wohops** [comment]
Migrated to Jira. New ticket number: JS-64
